File: core/src/main/java/org/kiwix/kiwixmobile/core/bookmark/BookmarksActivity.java
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.kiwix.kiwixmobile.core.bookmark;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageView;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.view.ActionMode;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import org.kiwix.kiwixmobile.core.Intents;
import org.kiwix.kiwixmobile.core.R;
import org.kiwix.kiwixmobile.core.R2;
import org.kiwix.kiwixmobile.core.base.BaseActivity;
import org.kiwix.kiwixmobile.core.extensions.ImageViewExtensionsKt;
import org.kiwix.kiwixmobile.core.main.CoreMainActivity;
import org.kiwix.kiwixmobile.core.zim_manager.ZimReaderContainer;

import static org.kiwix.kiwixmobile.core.utils.Constants.EXTRA_CHOSE_X_TITLE;
import static org.kiwix.kiwixmobile.core.utils.Constants.EXTRA_CHOSE_X_URL;

public class BookmarksActivity extends BaseActivity implements BookmarksContract.View,
  BookmarksAdapter.OnItemClickListener {

  private final List<BookmarkItem> bookmarksList = new ArrayList<>();
  private final List<BookmarkItem> allBookmarks = new ArrayList<>();
  private final List<BookmarkItem> deleteList = new ArrayList<>();

  @BindView(R2.id.toolbar)
  Toolbar toolbar;
  @BindView(R2.id.recycler_view)
  RecyclerView recyclerView;
  @Inject
  BookmarksContract.Presenter presenter;
  @Inject
  ZimReaderContainer zimReaderContainer;

  private boolean refreshAdapter = true;
  private BookmarksAdapter bookmarksAdapter;
  private ActionMode actionMode;

  private final ActionMode.Callback actionModeCallback = new ActionMode.Callback() {
    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
      mode.getMenuInflater().inflate(R.menu.menu_context_delete, menu);
      return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
      return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
      refreshAdapter = false;
      if (item.getItemId() == R.id.menu_context_delete) {
        allBookmarks.removeAll(deleteList);
        for (BookmarkItem bookmark : deleteList) {
          int position = bookmarksList.indexOf(bookmark);
          bookmarksList.remove(bookmark);
          bookmarksAdapter.notifyItemRemoved(position);
          bookmarksAdapter.notifyItemRangeChanged(position, bookmarksAdapter.getItemCount());
        }
        presenter.deleteBookmarks(new ArrayList<>(deleteList));
        mode.finish();
        return true;
      }
      return false;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
      if (deleteList.size() != 0) {
        deleteList.clear();
      }
      actionMode = null;
      if (refreshAdapter) {
        bookmarksAdapter.notifyDataSetChanged();
      }
    }
  };

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    presenter.attachView(this);
    setContentView(R.layout.activity_bookmarks);
    setSupportActionBar(toolbar);

    ActionBar actionBar = getSupportActionBar();
    if (actionBar != null) {
      actionBar.setDisplayHomeAsUpEnabled(true);
      actionBar.setTitle(R.string.menu_bookmarks);
    }

    bookmarksAdapter = new BookmarksAdapter(bookmarksList, deleteList, this);
    recyclerView.setAdapter(bookmarksAdapter);
  }

  @Override
  protected void onResume() {
    super.onResume();
    presenter.loadBookmarks(sharedPreferenceUtil.getShowBookmarksCurrentBook());
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.menu_bookmarks, menu);
    MenuItem toggle = menu.findItem(R.id.menu_bookmarks_toggle);
    toggle.setChecked(sharedPreferenceUtil.getShowBookmarksCurrentBook());

    SearchView search = (SearchView) menu.findItem(R.id.menu_bookmarks_search).getActionView();
    search.setQueryHint(getString(R.string.search_bookmarks));
    search.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
      @Override
      public boolean onQueryTextSubmit(String query) {
        return false;
      }

      @Override
      public boolean onQueryTextChange(String newText) {
        bookmarksList.clear();
        bookmarksList.addAll(allBookmarks);
        if ("".equals(newText)) {
          bookmarksAdapter.notifyDataSetChanged();
          return true;
        }
        presenter.filterBookmarks(bookmarksList, newText);
        return true;
      }
    });
    return true;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    int itemId = item.getItemId();
    if (itemId == android.R.id.home) {
      onBackPressed();
      return true;
    } else if (itemId == R.id.menu_bookmarks_toggle) {
      item.setChecked(!item.isChecked());
      sharedPreferenceUtil.setShowBookmarksCurrentBook(item.isChecked());
      presenter.loadBookmarks(sharedPreferenceUtil.getShowBookmarksCurrentBook());
      return true;
    }
    return super.onOptionsItemSelected(item);
  }

  @Override
  protected void onDestroy() {
    presenter.detachView();
    super.onDestroy();
  }

  @Override
  public void updateBookmarksList(List<BookmarkItem> bookmarksList) {
    allBookmarks.clear();
    allBookmarks.addAll(bookmarksList);
    notifyBookmarksListFiltered(bookmarksList);
  }

  @Override
  public void notifyBookmarksListFiltered(List<BookmarkItem> bookmarksList) {
    this.bookmarksList.clear();
    this.bookmarksList.addAll(bookmarksList);
    bookmarksAdapter.notifyDataSetChanged();
  }

  @Override
  public void onItemClick(ImageView favicon, BookmarkItem bookmark) {
    if (actionMode == null) {
      Intent intent = Intents.internal(CoreMainActivity.class);
      if ("null".equals(bookmark.getBookmarkUrl())) {
        intent.putExtra(EXTRA_CHOSE_X_TITLE, bookmark.getBookmarkTitle());
      } else {
        intent.putExtra(EXTRA_CHOSE_X_URL, bookmark.getBookmarkUrl());
      }
      if (bookmark.getZimFilePath() != null && !bookmark.getZimFilePath()
        .equals(zimReaderContainer.getZimCanonicalPath())) {
        intent.setData(Uri.fromFile(new File(bookmark.getZimFilePath())));
      }
      if (Settings.System.getInt(getContentResolver(), Settings.Global.ALWAYS_FINISH_ACTIVITIES, 0)
        == 1) {
        startActivity(intent);
      } else {
        setResult(RESULT_OK, intent);
      }
      finish();
    } else {
      toggleSelection(favicon, bookmark);
    }
  }

  @Override
  public boolean onItemLongClick(ImageView favicon, BookmarkItem bookmark) {
    if (actionMode != null) {
      return false;
    }
    actionMode = startSupportActionMode(actionModeCallback);
    refreshAdapter = true;
    toggleSelection(favicon, bookmark);
    return true;
  }

  private void toggleSelection(ImageView favicon, BookmarkItem bookmark) {
    if (deleteList.remove(bookmark)) {
      ImageViewExtensionsKt.setBitmapFromString(favicon, bookmark.getFavicon());
    } else {
      favicon.setImageDrawable(
        ContextCompat.getDrawable(this, R.drawable.ic_check_circle_blue_24dp));
      deleteList.add(bookmark);
    }
    actionMode.setTitle(getString(R.string.selected_items, deleteList.size()));

    if (deleteList.size() == 0) {
      actionMode.finish();
    }
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/bookmark/BookmarksActivity.kt
package org.kiwix.kiwixmobile.core.bookmark

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.view.ActionMode
import androidx.appcompat.view.ActionMode.Callback
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.activity_bookmarks.bookmarks_switch
import kotlinx.android.synthetic.main.activity_bookmarks.no_bookmarks
import kotlinx.android.synthetic.main.activity_history.recycler_view
import kotlinx.android.synthetic.main.layout_toolbar.toolbar
import org.kiwix.kiwixmobile.core.R
import org.kiwix.kiwixmobile.core.base.BaseActivity
import org.kiwix.kiwixmobile.core.bookmark.adapter.BookmarkDelegate.BookmarkItemDelegate
import org.kiwix.kiwixmobile.core.bookmark.adapter.BookmarkItem
import org.kiwix.kiwixmobile.core.bookmark.adapter.BookmarksAdapter
import org.kiwix.kiwixmobile.core.bookmark.adapter.BookmarksAdapter.OnItemClickListener
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.ExitActionModeMenu
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.ExitBookmarks
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.Filter
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.OnItemClick
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.OnItemLongClick
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.UserClickedDeleteButton
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.UserClickedDeleteSelectedBookmarks
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.Action.UserClickedShowAllToggle
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.BookmarkState
import org.kiwix.kiwixmobile.core.bookmark.viewmodel.BookmarkViewModel
import org.kiwix.kiwixmobile.core.di.components.CoreComponent
import org.kiwix.kiwixmobile.core.extensions.ActivityExtensions.coreActivityComponent
import org.kiwix.kiwixmobile.core.extensions.ActivityExtensions.viewModel
import org.kiwix.kiwixmobile.core.utils.SimpleTextListener
import javax.inject.Inject

class BookmarksActivity : OnItemClickListener, BaseActivity() {
  val activityComponent by lazy { coreActivityComponent }
  private var actionMode: ActionMode? = null
  @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
  private val bookmarksViewModel by lazy { viewModel<BookmarkViewModel>(viewModelFactory) }
  private val compositeDisposable = CompositeDisposable()

  private val actionModeCallback: Callback =
    object : Callback {
      override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        mode.menuInflater.inflate(R.menu.menu_context_delete, menu)
        return true
      }

      override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean = false

      override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_context_delete) {
          bookmarksViewModel.actions.offer(UserClickedDeleteSelectedBookmarks)
          return true
        }
        bookmarksViewModel.actions.offer(ExitActionModeMenu)
        return false
      }

      override fun onDestroyActionMode(mode: ActionMode) {
        bookmarksViewModel.actions.offer(ExitActionModeMenu)
        actionMode = null
      }
    }

  private val bookmarksAdapter: BookmarksAdapter by lazy {
    BookmarksAdapter(BookmarkItemDelegate(this))
  }

  override fun injection(coreComponent: CoreComponent) {
    activityComponent.inject(this)
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_bookmarks)
    setSupportActionBar(toolbar)

    supportActionBar?.setDisplayHomeAsUpEnabled(true)
    supportActionBar?.setTitle(R.string.bookmarks)

    recycler_view.adapter = bookmarksAdapter
    recycler_view.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)

    bookmarks_switch.setOnCheckedChangeListener { _, isChecked ->
      bookmarksViewModel.actions.offer(UserClickedShowAllToggle(isChecked))
    }
    bookmarks_switch.isChecked = sharedPreferenceUtil.showBookmarksAllBooks

    compositeDisposable.add(bookmarksViewModel.effects.subscribe { it.invokeWith(this) })
  }

  override fun onDestroy() {
    compositeDisposable.clear()
    super.onDestroy()
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.menu_bookmarks, menu)
    val search = menu.findItem(R.id.menu_bookmarks_search).actionView as SearchView
    search.queryHint = getString(R.string.search_bookmarks)
    search.setOnQueryTextListener(SimpleTextListener {
      bookmarksViewModel.actions.offer(Filter(it))
    })
    bookmarksViewModel.state.observe(this, Observer(::render))
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    if (item.itemId == android.R.id.home) {
      bookmarksViewModel.actions.offer(ExitBookmarks)
    }
    if (item.itemId == R.id.menu_bookmarks_clear) {
      bookmarksViewModel.actions.offer(UserClickedDeleteButton)
    }
    return super.onOptionsItemSelected(item)
  }

  private fun render(state: BookmarkState) {
    val filteredBookmarks = state.filteredBookmarks
    filteredBookmarks.let { bookmarksAdapter.items = it }
    toggleNoBookmarksText(filteredBookmarks)
    bookmarks_switch.isEnabled = !state.isInSelectionState
    if (state.isInSelectionState) {
      handleSelectionState(filteredBookmarks)
    } else {
      actionMode?.finish()
    }
  }

  private fun handleSelectionState(filteredBookmarks: List<BookmarkItem>) {
    if (actionMode == null) {
      actionMode = startSupportActionMode(actionModeCallback)
    }
    val numberOfSelectedItems = filteredBookmarks.filter(BookmarkItem::isSelected).size
    actionMode?.title = getString(R.string.selected_items, numberOfSelectedItems)
  }

  private fun toggleNoBookmarksText(items: List<BookmarkItem>) {
    if (items.isEmpty()) {
      no_bookmarks.visibility = View.VISIBLE
    } else {
      no_bookmarks.visibility = View.GONE
    }
  }

  override fun onItemClick(bookmark: BookmarkItem) {
    bookmarksViewModel.actions.offer(OnItemClick(bookmark))
  }

  override fun onItemLongClick(bookmark: BookmarkItem): Boolean =
    bookmarksViewModel.actions.offer(OnItemLongClick(bookmark))
}
